pkg/ipam/metadata/manager.go

```diff
@@ -113,17 +146,189 @@ func (m *manager) GetIPPoolForPod(owner string, family ipam.Family) (pool string
+		m.logger.Debug("pool selector: found pool by namespace annotation",
+			logfields.Owner, owner,
+			logfields.K8sNamespace, namespace,
+			logfields.PoolName, ippool,
+		)
 		return ippool, nil
 	}
 
-	// Fallback to default pool
-	return ipam.PoolDefault().String(), nil
+	podLabels := maps.Clone(pod.Labels)
+	if podLabels == nil {
+		podLabels = make(map[string]string)
+	}
+	// Add synthetic fields for selectors
+	podLabels[consts.PodNamespaceLabel] = namespace
+	podLabels[consts.PodNameLabel] = name
+
+	var matches []string
+	m.poolMu.RLock()
+	for _, cp := range m.compiledPools {
+		if family == ipam.IPv4 && !cp.hasV4 {
+			continue
+		}
+		if family == ipam.IPv6 && !cp.hasV6 {
+			continue
+		}
+		// Check if pod matches the pool's podSelector (if specified)
+		podMatches := cp.podSelector == nil || cp.podSelector.Matches(labels.Set(podLabels))
+
+		// Check if namespace matches the pool's namespaceSelector (if specified)
+		namespaceMatches := cp.namespaceSelector == nil || cp.namespaceSelector.Matches(labels.Set(podNamespace.Labels))
+
+		if podMatches && namespaceMatches {
+			matches = append(matches, cp.name)
+		}
+	}
+	m.poolMu.RUnlock()
+
+	switch len(matches) {
+	case 0:
+		// Check if pod or namespace requires pool match
+		podRequiresMatch := pod.Annotations[annotation.IPAMRequirePoolMatch] == "true"
+		namespaceRequiresMatch := podNamespace.Annotations[annotation.IPAMRequirePoolMatch] == "true"
+		if podRequiresMatch || namespaceRequiresMatch {
+			var annotationLocation string
+			if podRequiresMatch {
+				annotationLocation = "pod"
+			} else {
+				annotationLocation = "namespace"
+			}
+			return "", fmt.Errorf("no matching CiliumPodIPPool found for pod %s (family=%s) and require-pool-match annotation is set on %s", owner, family, annotationLocation)
+		}
+		// Fallback to default pool
+		m.logger.Debug("pool selector: no matches, falling back to default pool",
+			logfields.Owner, owner,
+		)
+		return ipam.PoolDefault().String(), nil
+	case 1:
+		m.logger.Debug("pool selector: found matching pool",
+			logfields.Owner, owner,
+			logfields.PoolName, matches[0],
+		)
+		return matches[0], nil
+	default:
+		// If multiple pools match, fail the allocation
+		m.logger.Error("pool selector: multiple pools matched; refusing to choose a pool",
+			logfields.Owner, owner,
+			logfields.Matches, matches,
+		)
+		return "", fmt.Errorf("multiple CiliumPodIPPools match pod %s (family=%s): %v", owner, family, matches)
+	}
+}
+
+// handlePoolEvent handles individual CiliumPodIPPool events and maintains internal selector state.
+func (m *manager) handlePoolEvent(ctx context.Context, event resource.Event[*cilium_v2alpha1.CiliumPodIPPool]) error {
+	defer func() {
+		event.Done(nil)
+	}()
+
+	switch event.Kind {
+	case resource.Sync:
+		m.logger.Debug("pool watcher: pools synced")
+		m.poolsSynced.Store(true)
+	case resource.Upsert:
+		_ = m.compilePool(event.Object)
```

Comment:
Considering `compilePool` already logs all errors it returns, and this is the only call site, let's remove the return value from `compilePool` since it's clearly not used.